ldms/src/ldmsd/ldmsd_config.c

```diff
@@ -173,64 +170,92 @@ ldmsd_plugin_t load_plugin(const char *plugin_name)
-	api_inst = calloc(1, sizeof (*api_inst) + api->base.context_size);
-	if (!api_inst)
-		return NULL;
-	memcpy(api_inst, api, sizeof(*api));
-	if (api->base.context_size)
-		api_inst->base.context = (void *)(api_inst + 1);
-	sampler = (void*)ldmsd_cfgobj_new_with_auth(cfg_name, LDMSD_CFGOBJ_SAMPLER,
-						sizeof(*sampler),
-						__del, uid, gid, perm);
-	if (!sampler) {
+	char log_name[4096];
+
+	sampler = (void *)ldmsd_cfgobj_new_with_auth(cfg_name,
+					     LDMSD_CFGOBJ_SAMPLER,
+					     sizeof(*sampler),
+					     __del, uid, gid, perm);
+	if (!sampler)
+	{
 		ovis_log(config_log, OVIS_LERROR,
-			"Error %d creating the sampler configuration object '%s'\n",
-				errno, cfg_name);
-		free(api_inst);
+			 "Error %d creating the sampler configuration object "
+			 "'%s'\n", errno, cfg_name);
 		return NULL;
 	}
-	sampler->api = api_inst;
-	sampler->api->base.cfg_name = strdup(cfg_name);
-	sampler->thread_id = -1;	/* stopped */
+	sprintf(log_name, "sampler.%s.%s", api->base.name, cfg_name);
+	sampler->log = ovis_log_register(log_name, "Sampler plugin log file.");
+	sampler->api = api;
+	sampler->libpath = strdup(libpath);
+	sampler->thread_id = -1; /* stopped */
+	if (sampler->api->base.constructor != NULL
+		&& sampler->api->base.destructor != NULL)
+	{
+		/* This plugin is multi-instance capable */
+		rc = sampler->api->base.constructor(&sampler->cfg);
+		if (rc)
+			goto err;
+	}
 #ifdef _CFG_REF_DUMP_
 	ref_dump(&sampler->cfg.ref, sampler->cfg.name, stderr);
 #endif
 	ldmsd_cfgobj_unlock(&sampler->cfg);
 	return sampler;
+err:
+#ifdef _CFG_REF_DUMP_
+	ref_dump(&sampler->cfg.ref, sampler->cfg.name, stderr);
+#endif
+	/* TODO: cleanup half created cfg obj */
+	errno = rc;
+	return NULL;
 }
 
-ldmsd_cfgobj_store_t ldmsd_store_add(const char *cfg_name,
-		struct ldmsd_store *api,
+ldmsd_cfgobj_store_t
+ldmsd_store_add(const char *cfg_name,
+		struct ldmsd_store *api, const char *libpath,
 		ldmsd_cfgobj_del_fn_t __del,
 		uid_t uid, gid_t gid, int perm)
 {
+	int rc;
 	ldmsd_cfgobj_store_t store;
-	struct ldmsd_store *api_inst;
-	api_inst = calloc(1, sizeof (*api_inst) + api->base.context_size);
-	if (!api_inst)
-		return NULL;
-	memcpy(api_inst, api, sizeof(*api));
-	if (api->base.context_size)
-		api_inst->base.context = (void *)(api_inst + 1);
+	char log_name[4096];
+
 	store = (void*)ldmsd_cfgobj_new_with_auth(cfg_name, LDMSD_CFGOBJ_STORE,
 						sizeof(*store),
 						__del, uid, gid, perm);
 	if (!store) {
 		ovis_log(config_log, OVIS_LERROR,
 			"Error %d creating the store configuration object '%s'\n",
-				errno, cfg_name);
-		free(api_inst);
+			 errno, cfg_name);
 		return NULL;
 	}
-	store->api = api_inst;
-	store->api->base.cfg_name = strdup(cfg_name);
+	sprintf(log_name, "store.%s.%s", api->base.name, cfg_name);
+	store->log = ovis_log_register(log_name, "Store plugin log file.");
+	store->api = api;
+	store->libpath = strdup(libpath);
+	if (store->api->base.constructor != NULL
+		&& store->api->base.destructor != NULL)
```

Comment:
I don't think we should have the "store->api->base.destructor != NULL" test in this conditional. If a constructor is implemented, but no destructor, it would be bad to just silently skip calling the constructor and continue as normal.

Granted it might be odd to have a constructor without a destructor. If we add a test that they both need to be present if one is present, that should really be a separate test with an error, probably earlier in the code.

But we're moving towards making most function calls optional. It is not immediately clear that we need to enforce that. At the moment I'm leaning towards not enforcing that they are a pair. But I could be convinced otherwise.

(Same change needed in the ldmsd_sampler_add)

Reply:
@morrone, you are arguing that the correct behavior is to crash when "bad-plugin" is loaded. I don't agree.

Reply:
Not at all. In fact crashing is exactly what this current code would lead to. I'm suggesting the way to _avoid_ a crash.

It sounds like you want to require that constructor and destructor are always present, or absent, as a pair. If that is what you want, then we need to have an explicit test for that. If the test fails, ldmsd should print an error message and then abort loading the plugin. Skipping a call that the plugin _has_ implemented is very likely to lead to a crash.